[wooki] step2 속성 변경 동작 구현 - 피드백 적용

banjjak2 opened the pull request:
다시 구현했습니다.. 이미지 뷰 대신 캔버스를 이용해 구현했습니다..

Q1. 3월 3일자 `다시 처음부터` 커밋부터 브랜치를 다시 만들고 싶은데 어떻게 해야될지 감이 안와서 하지 못했습니다. 따로 빼내는 방법이 있을까요?

Q2. 프레젠터에서 현재 선택한 Rectangle을 가지고 있어도 되는지 궁금합니다.

banjjak2 committed:
refact: Main 인터페이스 수정

LiveData 적용을 위해 뷰 인터페이스에 update 메소드 추가
생성 요청만 Plane에 하고 데이터는 프레젠터의 observe에서 받도록 하기 위해 create 메소드 수정

banjjak2 committed:
refact: LiveData 적용을 위한 Plane 객체 수정

LiveData 적용을 위해 private 타입 MutableLiveData 객체와 public 타입 LiveData 객체를 생성하여 외부에서는 get만 가능하고 내부에서는 get, set이 가능하도록 구현

banjjak2 committed:
refact: 프레젠터 옵저버 기능 추가

banjjak2 committed:
refact: 메인 화면 구현 수정 및 update 기능 추가

banjjak2 committed:
다시 처음부터...

banjjak2 committed:
feat: 사각형 그리는 커스텀 뷰 생성

banjjak2 committed:
style: 메인 액티비티에 커스텀 뷰 추가 및 레이아웃 depth 수정

banjjak2 committed:
feat: MVP 인터페이스 생성

banjjak2 committed:
refact: Presenter의 drawRectangle 반환값 삭제

굳이 rectangle 반환 없이 View의 showRectangle 메소드를 이용하면 됨

banjjak2 committed:
feat: Plane 객체 생성

banjjak2 committed:
feat: MainPresenter의 drawRectangle 구현

banjjak2 committed:
feat: 메인 액티비티의 showRectangle 구현

banjjak2 committed:
refact: Plane의 createRectangle 메소드 변경 및 전체 반환 메소드 추가

createRectangle에서 하나만 반환하면 Plane의 list와 drawview에서의 list가 일치하지 않을 수 있다고 판단해서 전체를 넘겨주는 방식으로 변경

banjjak2 committed:
refact: showRectangle 매개변수 변경 및 border전달시 null값 허용

빈 곳을 클릭한 경우 rectangle이 아닌 null을 반환해야 한다고 판단

banjjak2 committed:
refact: 프레젠터의 drawRectangle 메소드 변경

banjjak2 committed:
refact: 리스트가 DrawView에서 추가/삭제될 수 없으므로 List로 변경

banjjak2 committed:
refact: 매개변수 변경

banjjak2 committed:
feat: 사각형 테두리 객체 구현

banjjak2 committed:
feat: 인터페이스에 사각형 테두리 객체 관련 메소드 선언

banjjak2 committed:
feat: 사각형 테두리 관련 메소드 추가

banjjak2 committed:
fix: 사각형 클릭 시 테두리 설정이 잘 안되는 문제 해결

banjjak2 committed:
feat: 선택 시 테두리 및 사각형 정보 가져오는 기능 구현

banjjak2 committed:
feat: 커스텀 뷰 클릭시 좌표값 반환 이벤트 핸들러 선언

banjjak2 committed:
feat: 터치 이벤트 발생 시 좌표값 핸들러로 전송

banjjak2 committed:
feat: 사각형 테두리 그리는 기능 구현

banjjak2 committed:
feat: 메인에서 커스텀뷰의 이벤트 핸들러에 등록

banjjak2 committed:
feat: 사각형 정보 표현 구현

banjjak2 committed:
feat: 사각형 테두리 정보 표현 구현

banjjak2 committed:
refact: 불필요한 메소드 제거 및 메소드명 명확하게 변경

showRectangle을 통해 모든 리스트를 가져오므로 따로 정보를 가져오거나 프레젠터에 현재 선택된 사각형을 알고 있다면 알파값 세팅 시 인덱스가 필요 없다고 판단

banjjak2 committed:
feat: SeekBar를 이용해 투명도 설정 구현

ivybae commented on app/src/main/java/com/example/kotlin_drawingapp/MainActivity.kt:
```diff
@@ -1,8 +1,42 @@
+        drawView = findViewById(R.id.drawView)
+        val btnGenerateRectangle = findViewById<Button>(R.id.btn_generate_rect)
+        btnGenerateRectangle.setOnClickListener {
+            presenter.drawRectangle()
```
@banjjak2 하단의 showRectangle 메서드 구현과 비교해서 생각해볼게요~ 
view가 draw하는 것은 이해가 되는데 presenter가 draw를 하는 것은 약간 의아하죠? 이 메서드 구현을 이전 커밋에서 봤는데 수정할 수 있는 방법이 있어보여요

banjjak2 replied:
뷰에서 프레젠터에 Rectangle생성을 요청한다는 의미로 prefix를 create로 변경했습니다.

ivybae commented on app/src/main/java/com/example/kotlin_drawingapp/DrawView.kt:
```diff
@@ -7,8 +7,19 @@ import android.graphics.Paint
+import com.example.kotlin_drawingapp.model.RectangleBorder
 
 class DrawView(context: Context?, attrs: AttributeSet?) : View(context, attrs) {
+    interface SetOnDrawViewTouchListener {
```
@banjjak2 interface naming convention을 따로 찾아도 잘 안나오지만, Android API 에서 제공하는 interface 이름들과도 어울리지 않아보입니다. 다시 수정해볼까요?

banjjak2 replied:
View 크래스의 인터페이스 목록을 보니 On으로 시작하는 인터페이스가 있었고, 좀 더 의미를 명확하게 하기 위해 OnTouchEvent로 변경했습니다.
그런데 새로 정의한 인터페이스명이 기존에 정의된 OnTouchEvent와 똑같은 이름인데 이런 부분은 좀 불합리한 부분일까요?

ivybae replied:
@banjjak2 기존의 interface 명과 이름이 동일하면 import 하실 때 구분이 잘 안되서 불편하실수도 있긴 해요~ 기존 이름에서 Set만 지운것도 괜찮아보입니다~

ivybae commented on app/src/main/java/com/example/kotlin_drawingapp/MainPresenter.kt:
```diff
@@ -26,6 +30,13 @@ class MainPresenter(
 
     override fun setCurrentSelectedRectangleAlpha(alpha: Int) {
-
+        val _currentSelectedRectangle = currentSelectedRectangle
```
@banjjak2 메서드 내 지역변수 이름은 _로 시작하지 않습니다~

banjjak2 replied:
임시변수의 의미로 prefix에 tmp를 붙였습니다. 

혹시나 다른 부분에 문제가 있다면 바로 수정하도록 하겠습니다.
피드백 감사드립니다 !

ivybae reviewed (commented):
수고하셨습니다, 커멘트와 함께 아래 내용 확인해주세요!
**A1.** rebase 혹은 cherry-pick을 활용해보실 수 있을것 같습니다.
개인 레포지토리에서 연습을 해보실까요? 다음 PR에서는 개선해주세요~
처음에 이전에 피드백 드린 내용이 반영되어 있지 않아서 약간 이상하다고 생각하고 있었습니다.

**A2.** 어제 클래스에서 데이터를 관리하는 Repository 클래스를 만들어서 분리할 수도 있겠다고 말씀드렸죠? 가능하시다면 다음주에는 분리해보시는 것도 좋겠습니다~

banjjak2 committed:
refact: 1번째 피드백 적용

뷰는 프레젠터에 Rectangle을 만들라는 요청을 하는게 맞으므로 메서드명을 draw에서 create로 변경

banjjak2 committed:
refact: 2번째 피드백 적용

View의 인터페이스를 살펴본 결과 On~Listener로 구현되어 있음을 확인하고 인터페이스명을 SetOn~Listener에서 On~Listener로 변경

banjjak2 committed:
refact: 3번째 피드백 적용

언더바 제외하고 임시 변수 의미로 tmp를 prefix에 붙임
